File: 532-k-diff-pairs-in-an-array/k-diff-pairs-in-an-array.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        unordered_map<int, int> freq;

        for (auto x : nums) {
            freq[x]++;
        }

        int count = 0;
        for (auto x : freq) {

            if (k == 0) {
                if (x.second > 1) {
                    count++;
                }
            } else if (freq.find(x.first + k) != freq.end()) {
                count++;
            }
        }
        return count;
    }
};

File: 532-k-diff-pairs-in-an-array/k-diff-pairs-in-an-array.java
import java.util.*;

class Solution {
    public int findPairs(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        
        // Count frequency of each number
        for (int num : nums) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }
        
        int count = 0;
        
        for (Map.Entry<Integer, Integer> entry : freq.entrySet()) {
            int key = entry.getKey();
            int value = entry.getValue();
            
            if (k == 0) {
                if (value > 1) {
                    count++;
                }
            } else if (freq.containsKey(key + k)) {
                count++;
            }
        }
        
        return count;
    }
}
